packages/libs/src/sdk/services/Solana/programs/PaymentRouterClient/PaymentRouterClient.ts

```diff
@@ -0,0 +1,251 @@
+    const {
+      amount,
+      mint,
+      splits,
+      total,
+      contentId,
+      contentType,
+      blockNumber,
+      buyerUserId,
+      accessType
+    } = await parseParams(
+      'createPurchaseContentInstructions',
+      CreatePurchaseContentInstructionsSchema
+    )(params)
+    return [
+      this.createTransferInstruction({ amount, mint }),
+      this.createRouteInstruction({ splits, total, mint }),
+      this.createMemoInstruction({
+        contentId,
+        contentType,
+        blockNumber,
+        buyerUserId,
+        accessType
+      })
+    ]
+  }
+
+  /**
+   * Creates or gets the intermediate funds token account for the program.
+   * Only needs to be created once per mint.
+   * @see {@link https://github.com/solana-labs/solana-program-library/blob/d72289c79a04411c69a8bf1054f7156b6196f9b3/token/js/src/actions/getOrCreateAssociatedTokenAccount.ts getOrCreateAssociatedTokenAccount}
+   */
+  public async getOrCreateProgramTokenAccount(
+    params: GetOrCreateProgramTokenAccountRequest
+  ): Promise<Account> {
+    const args = await parseParams(
+      'getOrCreateProgramTokenAccount',
+      GetOrCreateProgramTokenAccountSchema
+    )(params)
+
+    // Check for cached account
+    const existingTokenAccount = this.existingTokenAccounts[args.mint]
+    if (existingTokenAccount) {
+      return existingTokenAccount
+    }
+
+    const mint = this.mints[args.mint]
+    if (!mint) {
+      throw new Error(`Mint ${args.mint} not configured`)
+    }
+    const associatedTokenAdddress = getAssociatedTokenAddressSync(
+      mint,
+      this.programAccount,
+      true
+    )
+
+    let account: Account | null = null
+    try {
+      account = await getAccount(this.connection, associatedTokenAdddress)
+      this.existingTokenAccounts[args.mint] = account
+    } catch (error: unknown) {
+      if (error instanceof TokenAccountNotFoundError) {
+        // As this isn't atomic, it's possible others can create associated accounts meanwhile.
+        try {
+          const instruction = createAssociatedTokenAccountIdempotentInstruction(
+            await this.getFeePayer(),
+            associatedTokenAdddress,
+            this.programAccount,
+            mint
+          )
+          const { lastValidBlockHeight, blockhash } =
+            await this.connection.getLatestBlockhash()
+          const msg = new TransactionMessage({
+            payerKey: await this.getFeePayer(),
+            recentBlockhash: blockhash,
+            instructions: [instruction]
+          })
+          const transaction = new VersionedTransaction(msg.compileToV0Message())
+          const signature = await this.sendTransaction(transaction)
+          await this.connection.confirmTransaction(
+            { signature, blockhash, lastValidBlockHeight },
+            'finalized'
+          )
+        } catch (e: unknown) {
+          // Ignore all errors; for now there is no API-compatible way to selectively ignore the expected
```

Comment:
Should we at least log them though?

Reply:
There's an expected error here if the account doesn't exist, and I figured it would be noisy. This code is a near copy/paste from @solana/web3.js but for creating a program-owned token account.